### ISSUE TITLE
example.py script does not cleanly shutdown producer / consumer instances

### ISSUE THREAD

**Issue by dakoller** [dpkp/kafka-python#1130]
Hi,

I took the example script with both producer & consumer.

The producer code looks like this:

----
class Producer(threading.Thread):
    daemon = True

    def run(self):
        producer = KafkaProducer(value_serializer=lambda v: json.dumps(v).encode('utf-8'))

        while True:
            producer.send('test',  {'msg': 'latency bad'})
            time.sleep(random.random() * 4)
            producer.send('test',  {'msg': 'showing better'})
            time.sleep(random.random()*4)

--
The output is:

Traceback (most recent call last):
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/threading.py", line 810, in __bootstrap_inner
    self.run()
  File "/Users/daniel/projects/kafka_message_processing/example.py", line 31, in run
    producer.send('test',  {'msg': 'latency bad'})
  File "/Library/Python/2.7/site-packages/kafka/producer/kafka.py", line 516, in send
    self.config['max_block_ms'])
  File "/Library/Python/2.7/site-packages/kafka/producer/record_accumulator.py", line 213, in append
    assert not self._closed, 'RecordAccumulator is closed'
AssertionError: RecordAccumulator is closed

...do you have any idea?

**Comment by harsh29** [comment]
+1

**Comment by dpkp** [comment]
Can you try enabling debug logs? It seems like something is closing your producer outside of the main thread. Possibly via interpreter shutdown? Do you only see this error when killing the script either via Ctrl-C or kill signal ?

**Comment by alirezastack** [comment]
I see the same state when trying to use the sample code on `OS X`. I do not kill the script using CTRL+C or any kill signal.

**Comment by dpkp** [comment]
Your debug logs suggest that you are closing the KafkaProducer while another thread is still trying to use the producer to send messages:
```
2017-09-17 22:16:58,075.75.5391120911:kafka.producer.kafka:140735279222784:INFO:5356:Closing the Kafka producer with 0 secs timeout.
```

**Comment by alirezastack** [comment]
I'm new to Kafka, and used the installation at my first step, so I don't have any other script in between that would try to close the producer connection.

**Comment by dpkp** [comment]
Ah -- sorry, I see that you are just running `example.py` . The threading code there does not handle clean shutdown, so the error is expected. It would be great if someone wanted to submit a PR that improves the example to avoid scary looking errors on close.

**Comment by TheAtomicOption** [comment]
I made the pull request above. Here's an expanded explanation of the changes:

Added better control over threads/processes, making them shutdown cleanly before the program terminates.

Switched from the iterator to poll() because using the iterator style will poll indefinitely looking for new messages when none are available. That prevents reaching any line receiving the exit signal, or executing any code to exit the thread cleanly. The only way I've found to get around that is to set consumer_time_ms which just forces a static length consumer lifetime (limiting), or to hold the producer open for a few seconds after the stop_event has been set and send one last message so that the iterator returns one last time allowing the for loop to execute the check for the stop_event (dirty).

The output is otherwise identical AFAICT.

Might be worth having parallel examples with documentation of these limitations. Or examples using assign() instead of subscribe() since I've not seen any assign() examples anywhere.
